File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/SecureMemory/Multithreaded/MultiThreadedSecretTest.cs
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.Utils;
using GoDaddy.Asherah.Logging;
using GoDaddy.Asherah.SecureMemory;
using GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl;
using Microsoft.Extensions.Logging;
using Xunit;

using static GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers.Constants;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.SecureMemory.Multithreaded
{
    public class MultiThreadedSecretTest
    {
        private static readonly ILogger Logger = LogManager.CreateLogger<MultiThreadedSecretTest>();
        private readonly byte[] payload;

        public MultiThreadedSecretTest()
        {
            payload = PayloadGenerator.CreateRandomBytePayload(PayloadSizeBytes);
        }

        [Fact]
        private void MultiThreadedWithSecretBytesAccess()
        {
            ISecretFactory secretFactory = new ProtectedMemorySecretFactory();
            Secret secret = secretFactory.CreateSecret(payload.Clone() as byte[]);

            // Get the current settings and try to force minWorkers
            ThreadPool.GetMinThreads(out _, out var currentMinIOC);
            Assert.True(ThreadPool.SetMinThreads(NumThreads, currentMinIOC));

            int completedTasks = 0;

            Parallel.ForEach(Enumerable.Range(0, NumThreads), i =>
            {
                try
                {
                    secret.WithSecretBytes(decryptedBytes =>
                    {
                        Assert.Equal(payload, decryptedBytes);
                        Interlocked.Increment(ref completedTasks);
                    });
                }
                catch (ThreadInterruptedException e)
                {
                    Logger.LogError(e, "Unexpected error during call");
                    throw;
                }
            });
            Assert.Equal(NumThreads, completedTasks);
        }
    }
}

File: tests/java/test-app/src/main/java/com/godaddy/asherah/testapp/securememory/multithreaded/MultiThreadedSecretTest.java
package com.godaddy.asherah.testapp.securememory.multithreaded;

import com.godaddy.asherah.securememory.Secret;
import com.godaddy.asherah.securememory.SecretFactory;
import com.godaddy.asherah.securememory.protectedmemoryimpl.ProtectedMemorySecretFactory;
import com.godaddy.asherah.testapp.ConfigurationParameterResolver;
import com.godaddy.asherah.testapp.ConfigurationParameterResolver.ConfigurationParameter;
import com.godaddy.asherah.testapp.utils.PayloadGenerator;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.LongAdder;
import java.util.stream.IntStream;

import static com.godaddy.asherah.testapp.testhelpers.Constants.*;
import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(ConfigurationParameterResolver.class)
class MultiThreadedSecretTest {
  private static final Logger LOG = LoggerFactory.getLogger(MultiThreadedSecretTest.class);

  private byte[] payload;

  @BeforeEach
  public void setupTest(@ConfigurationParameter(TEST_PARAM_PAYLOAD_SIZE_BYTES) final int payloadSizeBytes) {
    payload = PayloadGenerator.createRandomBytePayload(payloadSizeBytes);
  }

  @Test
  void multiThreadedWithSecretBytesAccess(
      @ConfigurationParameter(TEST_PARAM_NUM_REQUESTS) final int numRequests,
      @ConfigurationParameter(TEST_PARAM_NUM_THREADS) final int numThreads,
      @ConfigurationParameter(TEST_PARAM_THREAD_POOL_TIMEOUT_SECONDS) final long threadPoolTimeoutSeconds) {
    SecretFactory secretFactory = new ProtectedMemorySecretFactory();
    Secret secret = secretFactory.createSecret(payload.clone());

    ExecutorService pool = Executors.newFixedThreadPool(numThreads);
    LongAdder tasksCompleted = new LongAdder();
    IntStream.range(0, numRequests)
        .parallel()
        .forEach(i -> pool.submit(() -> {
          try {
            secret.withSecretBytes(decryptedBytes -> {
              assertArrayEquals(payload, decryptedBytes);
              tasksCompleted.increment();
            });
          }
          catch (Exception e) {
            LOG.error("unexpected error during call", e);
            throw e;
          }
        }));

    try {
      pool.shutdown();
      pool.awaitTermination(threadPoolTimeoutSeconds, TimeUnit.SECONDS);
    }
    catch (InterruptedException e) {
      LOG.error("interrupted exception during thread pool shutdown", e);
    }
    assertEquals(numRequests, tasksCompleted.sum());
  }
}
